presto-main/src/test/java/com/facebook/presto/sql/planner/optimizations/TestMergePartialAggregationsWithFilter.java

```diff
@@ -0,0 +1,73 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.sql.planner.optimizations;
+
+import com.facebook.presto.Session;
+import com.facebook.presto.spi.plan.AggregationNode;
+import com.facebook.presto.sql.planner.assertions.BasePlanTest;
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableMap;
+import org.testng.annotations.Test;
+
+import java.util.Optional;
+
+import static com.facebook.presto.SystemSessionProperties.MERGE_AGGS_WITH_AND_WITHOUT_FILTER;
+import static com.facebook.presto.SystemSessionProperties.PARTIAL_AGGREGATION_STRATEGY;
+import static com.facebook.presto.sql.planner.assertions.PlanMatchPattern.aggregation;
+import static com.facebook.presto.sql.planner.assertions.PlanMatchPattern.anyTree;
+import static com.facebook.presto.sql.planner.assertions.PlanMatchPattern.expression;
+import static com.facebook.presto.sql.planner.assertions.PlanMatchPattern.functionCall;
+import static com.facebook.presto.sql.planner.assertions.PlanMatchPattern.project;
+import static com.facebook.presto.sql.planner.assertions.PlanMatchPattern.singleGroupingSet;
+import static com.facebook.presto.sql.planner.assertions.PlanMatchPattern.tableScan;
+
+public class TestMergePartialAggregationsWithFilter
+        extends BasePlanTest
```

Comment:
Can you add some tests for the following (plan and correctness tests):
1. that the optimization doesn't trigger when disabled
2. multiple aggregations in the aggregation node
3. multiple levels of aggregation/aggregation nodes. (aggregating over a subquery that also has an aggregation in it).
4. non-grouped aggregation 
6. aggregations with order bys (either same order by and the aggregation triggers, or different order by and it doesn't)
7. aggregations with joins (how does this optimizer interact with the optimizer to push partial agg through join)
8. Can you have a query with grouping keys and grouping sets? Is that handled correctly?

Reply:
> 6. aggregations with joins (how does this optimizer interact with the optimizer to push partial agg through join)

According to my understanding, currently `PushPartialAggregationThroughJoin` is not applied in any cases, as this rule only works on partial aggregation node, however `PushPartialAggregationThroughExchange` is the optimizer which adds partial aggregation node, which is after the `PushPartialAggregationThroughJoin` optimizer.

Currently the optimizer rule will throw "PrestoException(GENERIC_INTERNAL_ERROR, "Unexpected plan node between partial and final aggregation")" if it sees join node between partial and final aggregation nodes.

Reply:
> 5. aggregations with order bys (either same order by and the aggregation triggers, or different order by and it doesn't)

Aggregations with group bys is considered as not decomposable and will not have partial aggregations, hence will not trigger the optimization. Added a test for it.